# <center>Тема 1. Первичный анализ данных с Pandas</center>
## <center>Практическое задание. Анализ данных пассажиров "Титаника"</center>

**Заполните код в клетках (где написано "Ваш код здесь") и ответьте на вопросы в [веб-форме](https://goo.gl/SbRcpW).**

In [0]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt

**Считаем данные из файла в память в виде объекта Pandas.DataFrame**

In [0]:
data = pd.read_csv('titanic_train.csv',
                  index_col='PassengerId')

**Данные представлены в виде таблицы. Посмотрим на первые 5 строк:**

In [21]:
data.head(5)

Survived  Pclass  ... Cabin Embarked
PassengerId                    ...               
1                   0       3  ...   NaN        S
2                   1       1  ...   C85        C
3                   1       3  ...   NaN        S
4                   1       1  ...  C123        S
5                   0       3  ...   NaN        S

[5 rows x 11 columns]

In [22]:
data.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

**Для примера отберем пассажиров, которые сели в Cherbourg (Embarked=C) и заплатили более 200 у.е. за билет (fare > 200).**

Убедитесь, что Вы понимаете, как эта конструкция работает. <br>
Если нет – посмотрите, как вычисляется выражение в квадратных в скобках.

In [23]:
data[(data['Embarked'] == 'C') & (data.Fare > 200)].head()

Survived  Pclass  ...            Cabin Embarked
PassengerId                    ...                          
119                 0       1  ...          B58 B60        C
259                 1       1  ...              NaN        C
300                 1       1  ...          B58 B60        C
312                 1       1  ...  B57 B59 B63 B66        C
378                 0       1  ...              C82        C

[5 rows x 11 columns]

**Можно отсортировать этих людей по убыванию платы за билет.**

In [24]:
data[(data['Embarked'] == 'C') & 
     (data['Fare'] > 200)].sort_values(by='Fare',
                               ascending=False).head()

Survived  Pclass  ...            Cabin Embarked
PassengerId                    ...                          
259                 1       1  ...              NaN        C
680                 1       1  ...      B51 B53 B55        C
738                 1       1  ...             B101        C
312                 1       1  ...  B57 B59 B63 B66        C
743                 1       1  ...  B57 B59 B63 B66        C

[5 rows x 11 columns]

**Пример создания признака.**

In [0]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    else:
        return 3

In [0]:
age_categories = [age_category(age) for age in data.Age]

In [0]:
data['Age_category'] = age_categories

**Другой способ – через `apply`.**

In [0]:
data['Age_category'] = data['Age'].apply(age_category)

**1. Сколько мужчин / женщин находилось на борту?**
- 412 мужчин и 479 женщин
- 314 мужчин и 577 женщин
- 479 мужчин и 412 женщин
- **577 мужчин и 314 женщин**

In [29]:
(data['Sex'] == 'male').sum(), (data['Sex'] == 'female').sum()

(577, 314)

**2. Выведите распределение переменной `Pclass` (социально-экономический статус) и это же распределение, только для мужчин / женщин по отдельности. Сколько было мужчин 2-го класса?**
- 104
- **108**
- 112
- 125

In [30]:
pd.crosstab(data['Pclass'], data['Sex'], margins=True)

Sex     female  male  All
Pclass                   
1           94   122  216
2           76   108  184
3          144   347  491
All        314   577  891

**3. Каковы медиана и стандартное отклонение платежей (`Fare`)? Округлите до 2 десятичных знаков.**
- **Медиана – 14.45, стандартное отклонение – 49.69**
- Медиана – 15.1, стандартное отклонение – 12.15
- Медиана – 13.15, стандартное отклонение – 35.3
- Медиана – 17.43, стандартное отклонение – 39.1

In [31]:
print("Медиана: ", round(data['Fare'].median(), 2))
print("Стандартное отклонение: ", round(data['Fare'].std(), 2))

Медиана:  14.45
Стандартное отклонение:  49.69


**4. Правда ли, что люди моложе 30 лет выживали чаще, чем люди старше 60 лет? Каковы доли выживших в обеих группах?**
- 22.7% среди молодых и 40.6% среди старых
- **40.6% среди молодых и 22.7% среди старых**
- 35.3% среди молодых и 27.4% среди старых
- 27.4% среди молодых и 35.3% среди старых

In [42]:
young_survived = data.loc[data['Age'] < 30, 'Survived']
old_survived = data.loc[data['Age'] > 60, 'Survived']

print("Shares of survived people: \n\t  among young {}%, \n\t  among old {}%.".format(
    round(100 * young_survived.mean(), 1), 
        round(100 * old_survived.mean(), 1)))

Shares of survived people: 
	  among young 40.6%, 
	  among old 22.7%.


**5. Правда ли, что женщины выживали чаще мужчин? Каковы доли выживших в обеих группах?**
- 30.2% среди мужчин и 46.2% среди женщин
- 35.7% среди мужчин и 74.2% среди женщин
- 21.1% среди мужчин и 46.2% среди женщин
- **18.9% среди мужчин и 74.2% среди женщин**

In [43]:
male_survived = data[data['Sex'] == 'male']['Survived']
female_survived = data[data['Sex'] == 'female']['Survived']


print("Shares of survived people: \n\t among women {}%, \n\t among men {}%".format(
    round(100 * female_survived.mean(), 1), round(100 * male_survived.mean(), 1)))

Shares of survived people: 
	 among women 74.2%, 
	 among men 18.9%


**6. Найдите самое популярное имя среди пассажиров Титаника мужского пола?**
- Charles
- Thomas
- **William**
- John

In [48]:
data['Name'].head()
data.loc[1, 'Name'].split(',')[1].split()[1]
first_names = data.loc[data['Sex'] == 'male', 'Name'].apply(lambda 
                                                            full_name: 
                full_name.split(',')[1].split()[1])
first_names.value_counts().head()

William    35
John       25
George     14
Charles    13
Thomas     13
Name: Name, dtype: int64

**7. Сравните графически распределение стоимости билетов и возраста у спасенных и у погибших. Средний возраст погибших выше, верно?**
- **Да**
- Нет

In [49]:
data.groupby('Survived')['Age'].mean()

Survived
0    30.626179
1    28.343690
Name: Age, dtype: float64

**8. Как отличается средний возраст мужчин / женщин в зависимости от класса обслуживания? Выберите верные утверждения:**
- **В среднем мужчины 1-го класса старше 40 лет**
- В среднем женщины 1-го класса старше 40 лет
- **Мужчины всех классов в среднем старше женщин того же класса**
- **В среднем люди в 1 классе старше, чем во 2-ом, а те старше представителей 3-го класса**

In [50]:
pd.crosstab(data['Pclass'], data['Sex'], 
            values=data['Age'], aggfunc=np.mean)

Sex        female       male
Pclass                      
1       34.611765  41.281386
2       28.722973  30.740707
3       21.750000  26.507589